kv_store/lww_kvs/kvs_memory_server.cpp

```diff
@@ -464,55 +509,248 @@ int main(int argc, char* argv[]) {
         for (auto set_iter = map_iter->second.begin(); set_iter != map_iter->second.end(); set_iter++) {
           communication::Key_Request_Tuple* tp = req.add_tuple();
           tp->set_key(*set_iter);
+          tp->set_global_memory_replication(placement[*set_iter].global_memory_replication_);
+          tp->set_global_ebs_replication(placement[*set_iter].global_ebs_replication_);
         }
         string key_req;
         req.SerializeToString(&key_req);
         zmq_util::send_string(key_req, &key_address_requesters[map_iter->first.key_exchange_connect_addr_]);
-        string target_address = zmq_util::recv_string(&key_address_requesters[map_iter->first.key_exchange_connect_addr_]);
-        (*res)[target_address] = unordered_set<string>(map_iter->second);
+        string key_res = zmq_util::recv_string(&key_address_requesters[map_iter->first.key_exchange_connect_addr_]);
+
+        communication::Key_Response resp;
+        resp.ParseFromString(key_res);
+
+        // for each key in the response
+        for (int i = 0; i < resp.tuple_size(); i++) {
+          string key = resp.tuple(i).key();
+          // we only have one address for memory tier
+          string target_address = resp.tuple(i).address(0).addr();
+          (*res)[target_address].insert(key);
+        }
+      }
+
+      // check if there are key requests to send to proxy
+      // NOTE: this is very similar to the code above, the only difference being the proxy address 
+      if (key_to_proxy.size() != 0) {
```

Comment:
It seems like we have a lot of code chunks that are repeated throughout the code. Can we factor these out into shared methods?